### PR TITLE
IFU mode improvements, continued

### PR DESCRIPTION
**Pull request by mperrin**
Continues IFU model improvements work started in #770. 

- [x] Adds a framework for, in IFU mode, after the initial PSF calculation completes calling a function to apply some transformation (e.g. a broadening convolution kernel) to improve the IFU model fidelity. 
- [x] For now, implements an initial placeholder version of that for NIRSpec and MIRI each as a simple Gaussian convolution kernel. This will be improved subsequently. 
- [x] Adds unit test functions for both NIRSpec and MIRI.
- [x] Minor improvements to logging outputs etc, incidental as part of testing and verifying the above

To Do:  _Note, these are deferred to future PRs, not in this one._

- [ ] Improved convolution kernel method for NIRSpec IFU
- [ ] Improved convolution kernel method for MIRI MRS, based on parameters from #691
- [ ] Further verification and validation compared to in-flight PSF calibrations.

### CODE REVIEW

**Comment by mperrin** [comment]
@patapisp FYI, I have started moving functionality from your PR #691 to this one for the MIRI MRS PSF models. I don't know if you might have some time to help check it, or to advise things that I should look at to check the result myself?  I know you're busy and time is short but let me know if there might be time we could chat on this soon. Cheers!

**Comment by patapisp** [comment]
@mperrin happy to check it out and test it.

File: webbpsf/detectors.py [hunk 1]
```diff
@@ -542,3 +547,129 @@ def _show_miri_cruciform_kernel(filt, npix=101, oversample=4, detector_position=
     ax.plot(0, 0, marker='+', color='yellow')
 
     matplotlib.pyplot.colorbar(mappable=ax.images[0])
+
+# Functions for applying IFU optics systematics models
+#
+# Note, thes are is not actually a "Detector" effect, but this file is a
```
Comment (marcio-melendez) [inline webbpsf/detectors.py]: "Thes are is not" --> This is not

File: webbpsf/detectors.py [hunk 2]
```diff
@@ -542,3 +547,129 @@ def _show_miri_cruciform_kernel(filt, npix=101, oversample=4, detector_position=
...
+    webbpsf.webbpsf_core._log.info(f'Applying MIRI IFU broadening model: {model_type}')
+    hdulist[ext].header.add_history(f"Added broadening model for IFU PSFs: {model_type}")
+
+    hdulist[ext].header['IFUBROAD'] = (True, "IFU PSF broadening model applied")
+    hdulist[ext].header['IFUBTYPE'] = (model_type, "IFU PSF broadening model type")
+
+    if model_type.lower() == 'gaussian':
+        # Very simple model just as a Gaussian convolution kernel
+        sigma = constants.INSTRUMENT_IFU_BROADENING_PARAMETERS['MIRI']['sigma']
+        hdulist[ext].header['IFUBSIGM'] = (model_type, "[arcsec] IFU PSF broadening Gaussian sigma")
```
Comment (marcio-melendez) [inline webbpsf/detectors.py]: For this header keyword, IFUBSIGM, it is sigma not model_type -> `(sigma, "[arcsec] IFU PSF broadening Gaussian sigma")`

File: webbpsf/detectors.py [hunk 3]
```diff
@@ -542,3 +547,129 @@ def _show_miri_cruciform_kernel(filt, npix=101, oversample=4, detector_position=
...
+                                   f'ext {hdulist[ext].header["EXTNAME"]}')
+
+    hdulist[ext].header['IFUBROAD'] = (True, "IFU PSF broadening model applied")
+    hdulist[ext].header['IFUBTYPE'] = (model_type, "IFU PSF broadening model type")
+    hdulist[ext].header.add_history(f"Added broadening model for IFU PSFs: {model_type}")
+
+    if model_type.lower() == 'gaussian':
+        sigma = constants.INSTRUMENT_IFU_BROADENING_PARAMETERS['NIRSPEC']['sigma']
+        # currently sigma= 50 mas, half a NIRSpec IFU spaxel. Approximate and loose estimate
+        hdulist[ext].header['IFUBSIGM'] = (model_type, "[arcsec] IFU PSF broadening Gaussian sigma")
```
Comment (marcio-melendez) [inline webbpsf/detectors.py]: It should be `sigma`  instead of `model_type`

**Review by marcio-melendez — COMMENTED** [review]
Just a minor typo in the description and a couple of minor header information fixes.  Everything else looks good, I tried the gaussian, empirical and None. Results are as expected including information inside the fits header

**Comment by mperrin** [comment]
@obi-wan76 Thanks for the review, and catching the couple little things to fix. All fixed and pushed now.

**Comment by marcio-melendez** [comment]
> @obi-wan76 Thanks for the review, and catching the couple little things to fix. All fixed and pushed now.

@mperrin I don't see the changes, did you push your latest fix?

File: webbpsf/detectors.py [hunk 4]
```diff
@@ -542,3 +547,129 @@ def _show_miri_cruciform_kernel(filt, npix=101, oversample=4, detector_position=
...
+    webbpsf.webbpsf_core._log.info(f'Applying MIRI IFU broadening model: {model_type}')
+    hdulist[ext].header.add_history(f"Added broadening model for IFU PSFs: {model_type}")
+
+    hdulist[ext].header['IFUBROAD'] = (True, "IFU PSF broadening model applied")
+    hdulist[ext].header['IFUBTYPE'] = (model_type, "IFU PSF broadening model type")
+
+    if model_type.lower() == 'gaussian':
+        # Very simple model just as a Gaussian convolution kernel
+        sigma = constants.INSTRUMENT_IFU_BROADENING_PARAMETERS['MIRI']['sigma']
+        hdulist[ext].header['IFUBSIGM'] = (sigma, "[arcsec] IFU PSF broadening Gaussian sigma")
```
Comment (marcio-melendez) [inline webbpsf/detectors.py:589]: I see the changes, thanks!

**Review by marcio-melendez — APPROVED** [review]

**Pull request merged**